client/lib/load-script/callback-handler.js

```diff
@@ -0,0 +1,96 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { includes } from 'lodash';
+import debugFactory from 'debug';
+const debug = debugFactory( 'lib/load-script/callback-handler' );
+
+/**
+ * Module variables
+ */
+const callbacksForURLsInProgress = new Map();
+
+export function getCallbacksMap() {
+	return callbacksForURLsInProgress;
+}
+
+export function isLoading( url ) {
+	return getCallbacksMap().has( url );
+}
+
+export function addScriptCallback( url, callback ) {
+	const callbacksMap = getCallbacksMap();
+	if ( isLoading( url ) ) {
+		debug( `Adding a callback for an existing script from "${ url }"` );
+		callbacksMap.get( url ).push( callback );
+	} else {
+		debug( `Adding a callback for a new script from "${ url }"` );
+		callbacksMap.set( url, [ callback ] );
+	}
+}
+
+export function removeScriptCallback( url, callback ) {
+	debug( `Removing a known callback for a script from "${ url }"` );
+
+	if ( ! isLoading( url ) ) {
+		return;
+	}
+
+	const callbacksMap = getCallbacksMap();
+	const callbacksAtUrl = callbacksMap.get( url );
+
+	if ( ! includes( callbacksAtUrl, callback ) ) {
+		return;
+	}
+
+	if ( 1 === callbacksAtUrl.length ) {
+		callbacksMap.delete( url );
+		return;
+	}
+
+	const index = callbacksAtUrl.indexOf( callback );
+	callbacksAtUrl.splice( index, 1 );
```

Comment:
just a note that we are at two levels of set membership: the URL and the list of callbacks.

any reason not to go full inception here and keep a set of sets?

```js
const urlCallbacks = new Map();

const addCallback = ( url, callback ) => {
	if ( ! urlCallbacks.has( url ) ) {
		urlCallbacks.set( url, new Set() );
	}

	urlCallbacks.get( url ).add( callback );
}

const removeCallback = ( url, callback ) => {
	if ( ! urlCallbacks.has( url ) ) {
		return;
	}

	urlCallbacks.get( url ).delete( callback );
}
```

as a bonus of using these data structures we can remove the verification logic as it's always safe to delete an element from a set even if it's not in the set already, etc…

We could go a step further if we needed/wanted of course and provide an updater for the mapped sets

```js
const updateCallbacks = ( url, updater ) => {
	if ( ! urlCallbacks.gas( url ) ) {
		urlCallbacks.set( url, new Set() );
	}

	updater( urlCallbacks.get( url ) );
}

const addCallback = ( url, callback ) =>
	updateMapSet( url, set => set.add( callback ) );

const removeCallback = ( url, callback ) =>
	updateMapSet( url, set => set.delete( callback ) );
```

Now we need no logic to check for the existence of the set at the url. Just some thoughts.

Reply:
Using a set here sounds like a fine idea, since we no longer need to check if it's safe to delete.

Using the updater proved to be a bit more troublesome than it was worth (from how I was going about it), so I punted it for now.